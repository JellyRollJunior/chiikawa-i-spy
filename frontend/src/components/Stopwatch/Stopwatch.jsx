import { useEffect, useState } from 'react';
import shared from '../../styles/shared.module.css';

const Stopwatch = ({ startTime, isRunning }) => {
  const [stopwatch, setStopwatch] = useState({
    minutes: 0,
    seconds: 0,
    centiseconds: 0,
  });

  useEffect(() => {
    if (!startTime || !isRunning) return;

    const intervalId = setInterval(() => {
      if (startTime && isRunning) {
        const start = new Date(startTime);
        const duration = new Date() - start;

        let minutes = Math.floor(duration / 60000);
        let seconds = Math.floor(duration / 1000) % 60;
        let centiseconds = Math.floor(duration / 10) % 100;

        if (minutes < 10) minutes = `0${minutes}`;
        if (seconds < 10) seconds = `0${seconds}`;
        if (centiseconds < 10) centiseconds = `0${centiseconds}`;
        setStopwatch({ minutes, seconds, centiseconds });
      }
    }, 10);

    return () => clearInterval(intervalId)
  }, [startTime, isRunning]);

  return (
    <h3 className={shared.primaryButton}>
      {stopwatch.minutes}:{stopwatch.seconds}:{stopwatch.centiseconds}
    </h3>
  );
};

export { Stopwatch };
