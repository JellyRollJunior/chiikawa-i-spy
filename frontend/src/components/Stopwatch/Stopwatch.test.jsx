import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { Stopwatch } from './Stopwatch.jsx';

describe('Stopwatch component', () => {
  it('Renders time in format mm:ss:cs', () => {
    const now = new Date();
    const { container } = render(
      <Stopwatch startTime={now} isRunning={false} />
    );

    expect(container).toMatchSnapshot();
  });
});
