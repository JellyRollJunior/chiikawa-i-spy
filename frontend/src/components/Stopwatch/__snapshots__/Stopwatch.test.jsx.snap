// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stopwatch component > Renders time in format mm:ss:cs 1`] = `
<div>
  <h3
    class="_primaryButton_5890fc"
  >
    0
    :
    0
    :
    0
  </h3>
</div>
`;
